x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java

```diff
@@ -141,17 +143,47 @@ private Object unwrapMultiValue(Object values) {
 
     @SuppressWarnings("unchecked")
     Object extractFromSource(Map<String, Object> map) {
-        Object value = map;
-        boolean first = true;
-        // each node is a key inside the map
-        for (String node : path) {
-            if (value == null) {
-                return null;
-            } else if (first || value instanceof Map) {
-                first = false;
-                value = ((Map<String, Object>) value).get(node);
-            } else {
-                throw new SqlIllegalArgumentException("Cannot extract value [{}] from source", fieldName);
+        Object value = null;
+
+        // Used to avoid recursive method calls
+        // Holds the sub-maps in the document hierarchy that are pending to be inspected.
+        LinkedList<Map<String, Object>> queue = new LinkedList<>();
```

Comment:
minor nitpick - use the `Deque` instead of `LinkedList`.

Reply:
Since the index and the Map are associated, how about using only one `Deque` which holds a `Tuple` instead of two `Deque`:

```
Deque<Tuple<Map<String, Object> index>>` queue = ...

if (node instanceof Map) {
   queue.add(new Tuple<>(node, Integer.valueOf(i));
}

```

Reply:
Actually, I had those 2 before while implementing. :-)

I chose `LinkedList` as we don't need any special operations of queue (we don't care about the order of dequeuing). I've just read about the better performance of `ArrayDequeue` over `LinkedList` and I'll change.

For the `Tuple` I had it as you said but I changed to a separate structure that holds integers to avoid the instantiation of `Tuple` objects on top of the Integer boxing. What do you think?

Reply:
And after benchmarking: https://gist.github.com/matriv/07e5909a49bed9794f350f458a0b2c60
The results show:
```
Benchmark                      Mode  Cnt       Score       Error  Units
MyBenchmark.testArrayDequeue  thrpt   25  608384,000 ± 7893,014  ops/s
MyBenchmark.testLinkedList    thrpt   25  581316,884 ±  8028,331  ops/s
```

Reply:
Autoboxing already happens and I wouldn't worry to much about it considering the depth is not that big. Same for `Linked` vs `Array` (in general arrays are faster except for inserting in the middle as that requires resizing/copying at which the linked structure excels).
I think the `Tuple` makes the code a bit more compact and safe (the queues cannot get out of sync) and more readable/simple code always trumps optimization (especially micro ones as here).

Reply:
Just FYI regarding the Tuples:
https://gist.github.com/matriv/3d2557ae1621e99a3c9505b5e4e6998f
and the result:
```
Benchmark                      Mode  Cnt       Score      Error  Units
MyBenchmark.testNoTuples      thrpt   25  601543,170 ± 7859,704  ops/s
MyBenchmark.testTuples        thrpt   25  573953,828 ± 5987,749  ops/s
```